File: node_examples/acBatch.js
/**
 * sample for connect multiple BLE devices(use batch-connect), and write data to all device
 * use Cassia AC APIs, refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki
 * to run the code, you should have a Cassia Router connected to a Cassia AC
 */
const request = require('request');
const EventSource = require('eventsource');
const qs = require('querystring');

/*
 * replace it with your AC address
 */
const AC_HOST = 'http://q1.lunxue.cc/api';

/*
 * you can set your developer key and secret under AC -> Settings -> Developer account for RESTful APIs
 */
const DEVELOPER_KEY = 'cassia';
const DEVELOPER_SECRET = 'cassia';

/*
 * this is your router's MAC, you should add the router to AC's online list first
 */
const ROUTER_MAC = 'CC:1B:E0:E0:28:EC';

const isWritten = {};

// convert http request to promise
function req(options) {
  return new Promise((resolve, reject) => {
    request(options, function (error, response) {
      if (error) reject(error);
      else if (response.statusCode !== 200) reject(response.body);
      else resolve(response.body);
    });
  });
}
/*
 * auth with your AC, the token will expired after ONE hour,
 * if you run a long term task, you should refresh the token every hour
 * refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/Getting-Started#access-cassia-router-through-the-cassia-ac
 */ 
function auth(key, secret) {
  let options = {
    method: 'POST',
    url: `${AC_HOST}/oauth2/token`,
    headers: {
      'Content-Type': 'application/json'
    },
    json: true,
    /*
     * auth options will encode Authorization header for you,
     * you can also add header "Authorization: Basic <Base64_encode(key + ':' + secret)>" manually
     */
    auth: {
      user: key,
      pass: secret,
    },
    body: {grant_type: 'client_credentials'}
  };
  return req(options);
}

/*
 * scan and connect devices
 * refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#scan-bluetooth-devices
 * Sever-Sent Event(SSE) is used in scan, connection-state and notify of Cassia RESTful API,
 * SSE spec: https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface
 * API will send ':keep-alive' every 30 seconds in SSE connection for user to check if the connection is active or not.
 * User need to call Cassia RESTful API to reconnect SSE in case that the connection is termincated abnormally, such as keep-alive lost, socket error, network problem, etc.
 * Nodejs library 'eventsource' handle the SSE reconnection automatically. For other lanuages, the reconnection may needs to be handled by users application.
 */
function openScanSseAndConnect(routerMac, token) {
  const query = {
    /*
     * filter devices whose rssi is below -75, and name begins with 'Cassia',
     * there are many other filters, you can find them in document
     * use proper filters can significantly reduce traffic between Router and AC
     */
    filter_rssi: -75,
    filter_name: 'Cassia*',
    /*
     * use active scan, default is passive scan
     * active scan makes devices response with data packet which usually contains device's name
     */
    active: 1,
    mac: routerMac, // which router you want to start scan
    access_token: token // you can put token in query 'access_token=<token>' or in header 'Bearer <token>' 
  };
  const url = `${AC_HOST}/gap/nodes?event=1&${qs.encode(query)}`;
  const sse = new EventSource(url);

  sse.on('error', function(error) {
    console.log('open notify sse failed:', error);
  });
  
  sse.on('message', function(message) {
    const data = JSON.parse(message.data);
    const deviceAddr = data.bdaddrs[0];
    const deviceMac = deviceAddr.bdaddr;
    const deviceAddrType = deviceAddr.bdaddrType;
    if (isWritten[deviceMac]) return;
    batchConnect(token, deviceMac, deviceAddrType);
  });
  
  return Promise.resolve(sse);
}

/*
 * Batch Connect/Disconnect to a Target Device (v2.0 and above)
 * refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#batch-connectdisconnect-to-a-target-device-v20-and-above
 */
function batchConnect(token, deviceMac, addrType) {
  let options = {
    'method': 'POST',
    'url': `${AC_HOST}/api/gap/batch-connect?mac=${routerMac}&access_token=${token}`,
    'headers': {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      /*
      * you can put multiple devices in list parameter, all the devices will enter a connecting queue
      * when device is connected, it will report connected event in connection-state SSE
      */
      list: [{"type": addrType,"addr": deviceMac}],
      timeout: 5000,
      per_dev_timeout: 10000
    })
  };

  return req(options);
}

/*
 * Read/Write the Value of a Specific Characteristic
 * refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#readwrite-the-value-of-a-specific-characteristic
 */
function write(token, deviceMac, handle, value) {
  let options = {
    method: 'GET',
    url: `${AC_HOST}/gatt/nodes/${deviceMac}/handle/${handle}/value/${value}?mac=${ROUTER_MAC}&access_token=${token}`,
  };
  return req(options);
}

/*
 * disconnect device
 * refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#connectdisconnect-to-a-target-device
 */
function disconnect(token, deviceMac) {
  let options = {
    'method': 'DELETE',
    'url': `${AC_HOST}/api/gap/nodes/${deviceMac}/connection?mac=${ROUTER_MAC}&access_token=${token}`,
  };
  return req(options);
}

/*
 * Get Device Connection Status
 * you will get notify when device connected or disconnected to a Router
 * refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#get-device-connection-status
 * Sever-Sent Event(SSE) is used in scan, connection-state and notify of Cassia RESTful API,
 * SSE spec: https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface
 * API will send ':keep-alive' every 30 seconds in SSE connection for user to check if the connection is active or not.
 * User need to call Cassia RESTful API to reconnect SSE in case that the connection is termincated abnormally, such as keep-alive lost, socket error, network problem, etc.
 * Nodejs library 'eventsource' handle the SSE reconnection automatically. For other lanuages, the reconnection may needs to be handled by users application.
 */
function openConnectStateSse(token) {
  const url = `${AC_HOST}/management/nodes/connection-state?mac=${routerMac}&access_token=${token}`;
  const sse = new EventSource(url);

  sse.on('error',function(error) {
    console.log('open connect status sse failed:', error);
  });
  
  sse.on('message', async function(message) {
    const data = JSON.parse(message.data);
    const deviceMac = data.handle;
    // skip device which have already written data
    if (isWritten[deviceMac]) return;
    if (data.connectionState === 'connected') {
      await write(token, deviceMac, 39, '21ff310302ff31');
      isWritten[deviceMac] = true;
      await disconnect(token, deviceMac);
    } else {
      // if device is disconnected, it can be scanned again
    }
  });
  
  return Promise.resolve(sse);
}

(async () => {
  try {
    let token = await auth(DEVELOPER_KEY, DEVELOPER_SECRET);
    console.log('token:', token);
    await openScanSseAndConnect(token);
    await openConnectStateSse(token);
    console.log('success');
  } catch(ex) {
    console.error('fail:', ex);
  }
})();

File: python_examples/acBatch.py
import asyncio
import aiohttp
from aiohttp_sse_client import client as sse_client_async
import json

# replace it with your AC address base URL
AC_BASE_URL = 'http://10.100.144.168'

# you can set your developer key and secret under AC -> Settings -> Developer account for RESTful APIs
DEVELOPER_KEY = 'admin1'
DEVELOPER_SECRET = '1q2w#E$R'

# this is your router's MAC, you should add the router to AC's online list first
ROUTER_MAC = 'CC:1B:E0:E2:E9:B8'

# this is your device scan filter
FILTER_MAC = 'CC:0A:19:32:6A:0A'

AC_HOST = f'{AC_BASE_URL}/api'

is_written = {}

async def req(session, method, url, headers=None, json_data=None, auth=None):
    async with session.request(method, url, headers=headers, json=json_data, auth=auth) as resp:
        try:
            if resp.content_type == 'application/json':
                return await resp.json()
            else:
                return await resp.text()
        except Exception as e:
            raise Exception(f"Failed to parse response: {e}")

# auth with your AC, the token will expired after ONE hour,
# if you run a long term task, you should refresh the token every hour
# refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/Getting-Started#access-cassia-router-through-the-cassia-ac
async def auth(session, key, secret):
    url = f"{AC_HOST}/oauth2/token"
    headers = {'Content-Type': 'application/json'}
    auth = aiohttp.BasicAuth(key, secret)
    data = {"grant_type": "client_credentials"}
    return await req(session, 'POST', url, headers=headers, json_data=data, auth=auth)

# scan and connect devices
# refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#scan-bluetooth-devices
async def open_scan_sse_and_connect(session, token):
    params = {
        'filter_rssi': -75,
        'filter_mac': FILTER_MAC,
        'active': 1,
        'mac': ROUTER_MAC,
        'access_token': token,
        'event': 1
    }
    url = f"{AC_HOST}/gap/nodes"
    
    async with sse_client_async.EventSource(url, session=session, params=params) as evts:
        async for event in evts:
            if event.data:
                data = json.loads(event.data)
                device_addr = data['bdaddrs'][0]
                device_mac = device_addr['bdaddr']
                device_addr_type = device_addr['bdaddrType']
                if device_mac not in is_written:
                    await batch_connect(session, token, device_mac, device_addr_type)

# Batch Connect/Disconnect to a Target Device (v2.0 and above)
# refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#batch-connectdisconnect-to-a-target-device-v20-and-above
async def batch_connect(session, token, device_mac, addr_type):
    url = f"{AC_HOST}/gap/batch-connect?mac={ROUTER_MAC}&access_token={token}"
    headers = {'Content-Type': 'application/json'}
    data = {
        "list": [{"type": addr_type, "addr": device_mac}],
        "timeout": 5000,
        "per_dev_timeout": 10000
    }
    await req(session, 'POST', url, headers=headers, json_data=data)

# Read/Write the Value of a Specific Characteristic
# refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#readwrite-the-value-of-a-specific-characteristic
async def write(session, token, device_mac, handle, value):
    url = f"{AC_HOST}/gatt/nodes/{device_mac}/handle/{handle}/value/{value}?mac={ROUTER_MAC}&access_token={token}"
    await req(session, 'GET', url)
    print('write ok:', device_mac)

# disconnect device
# refer: https://github.com/CassiaNetworks/CassiaSDKGuide/wiki/RESTful-API#connectdisconnect-to-a-target-device
async def disconnect(session, token, device_mac):
    url = f"{AC_HOST}/gap/nodes/{device_mac}/connection?mac={ROUTER_MAC}&access_token={token}"
    await req(session, 'DELETE', url)
    print('disconnect ok:', device_mac)

# Get Device Connection Status
async def open_connect_state_sse(session, token):
    url = f"{AC_HOST}/management/nodes/connection-state?mac={ROUTER_MAC}&access_token={token}"
    
    async with sse_client_async.EventSource(url, session=session) as evts:
        async for event in evts:
            if event.data:
                print(event.data)
                data = json.loads(event.data)
                device_mac = data['handle']
                if device_mac not in is_written:
                    if data['connectionState'] == 'connected':
                        await write(session, token, device_mac, 24, '0100')
                        is_written[device_mac] = True
                        # TODO: remove batch connect
                        await disconnect(session, token, device_mac)

async def main():
    timeout = aiohttp.ClientTimeout(total=None)
    async with aiohttp.ClientSession(timeout=timeout) as session:
        try:
            token_data = await auth(session, DEVELOPER_KEY, DEVELOPER_SECRET)
            print('token:', token_data)
            token = token_data['access_token']
            
            # Run both SSE connections concurrently
            await asyncio.gather(
                open_scan_sse_and_connect(session, token),
                open_connect_state_sse(session, token)
            )
            print('success')
        except Exception as ex:
            print('fail:', ex)

if __name__ == "__main__":
    asyncio.run(main())